rpc tests leak clients

The following test appears to have failed:

[#4379](https://circleci.com/gh/cockroachdb/cockroach/4379):

```
=== RUN TestUnretryableError
I0701 15:05:50.076790     131 rpc/client.go:121] client 127.0.0.1:45109 connected
--- PASS: TestUnretryableError (0.05s)
=== RUN TestClientNotReady
E0701 15:05:50.124186     131 rpc/client.go:118] client 127.0.0.1:55595 failed to connect: client is closed
--- FAIL: TestClientNotReady (0.14s)
	leaktest.go:136: Test appears to have leaked an rpc client:
		net.(*pollDesc).Wait(0xc2080e3800, 0x72, 0x0, 0x0)
			/usr/src/go/src/net/fd_poll_runtime.go:84 +0x47
		net.(*pollDesc).WaitRead(0xc2080e3800, 0x0, 0x0)
			/usr/src/go/src/net/fd_poll_runtime.go:89 +0x43
		net.(*netFD).Read(0xc2080e37a0, 0xc208370800, 0x800, 0x800, 0x0, 0x7fb4c4aa9dc0, 0xc2084a0018)
			/usr/src/go/src/net/fd_unix.go:242 +0x40f
		net.(*conn).Read(0xc20802e168, 0xc208370800, 0x800, 0x800, 0x0, 0x0, 0x0)
			/usr/src/go/src/net/net.go:121 +0xdc
		crypto/tls.(*block).readFromUntil(0xc2084484e0, 0x7fb4c4aac098, 0xc20802e168, 0x5, 0x0, 0x0)
--
E0701 15:05:50.272553     131 rpc/client.go:118] client 127.0.0.1:60196 failed to connect: rpc/client.go:178: system is stopping
E0701 15:05:50.272657     131 rpc/client.go:118] client 127.0.0.1:60344 failed to connect: rpc/client.go:178: system is stopping
E0701 15:05:50.272725     131 rpc/client.go:118] client 127.0.0.1:43545 failed to connect: rpc/client.go:178: system is stopping
E0701 15:05:50.272830     131 rpc/client.go:118] client 127.0.0.1:53609 failed to connect: rpc/client.go:178: system is stopping
E0701 15:05:50.287610     131 rpc/client.go:118] client 127.0.0.1:47417 failed to connect: unexpected EOF
--- FAIL: TestComplexScenarios (0.16s)
	leaktest.go:136: Test appears to have leaked an rpc client:
		net.(*pollDesc).Wait(0xc2080e3800, 0x72, 0x0, 0x0)
			/usr/src/go/src/net/fd_poll_runtime.go:84 +0x47
		net.(*pollDesc).WaitRead(0xc2080e3800, 0x0, 0x0)
			/usr/src/go/src/net/fd_poll_runtime.go:89 +0x43
		net.(*netFD).Read(0xc2080e37a0, 0xc208370800, 0x800, 0x800, 0x0, 0x7fb4c4aa9dc0, 0xc2084a0018)
			/usr/src/go/src/net/fd_unix.go:242 +0x40f
		net.(*conn).Read(0xc20802e168, 0xc208370800, 0x800, 0x800, 0x0, 0x0, 0x0)
			/usr/src/go/src/net/net.go:121 +0xdc
		crypto/tls.(*block).readFromUntil(0xc2084484e0, 0x7fb4c4aac098, 0xc20802e168, 0x5, 0x0, 0x0)
--
			github.com/cockroachdb/cockroach/rpc/_test/_testmain.go:104 +0x1d1
=== RUN TestUpdatedAddr
W0701 15:05:50.379749     131 rpc/server.go:216] asked for port 1234, got 1235
W0701 15:05:50.379796     131 rpc/server.go:216] asked for port 1234, got 1235
W0701 15:05:50.379836     131 rpc/server.go:216] asked for port 1234, got 1235
--- FAIL: TestUpdatedAddr (0.09s)
	leaktest.go:136: Test appears to have leaked an rpc client:
		net.(*pollDesc).Wait(0xc2080e3800, 0x72, 0x0, 0x0)
			/usr/src/go/src/net/fd_poll_runtime.go:84 +0x47
		net.(*pollDesc).WaitRead(0xc2080e3800, 0x0, 0x0)
			/usr/src/go/src/net/fd_poll_runtime.go:89 +0x43
		net.(*netFD).Read(0xc2080e37a0, 0xc208370800, 0x800, 0x800, 0x0, 0x7fb4c4aa9dc0, 0xc2084a0018)
			/usr/src/go/src/net/fd_unix.go:242 +0x40f
		net.(*conn).Read(0xc20802e168, 0xc208370800, 0x800, 0x800, 0x0, 0x0, 0x0)
			/usr/src/go/src/net/net.go:121 +0xdc
		crypto/tls.(*block).readFromUntil(0xc2084484e0, 0x7fb4c4aac098, 0xc20802e168, 0x5, 0x0, 0x0)
--
			/go/src/github.com/cockroachdb/cockroach/util/leaktest/leaktest.go:34 +0x28
		github.com/cockroachdb/cockroach/rpc.TestMain(0xc20800a910)
			/go/src/github.com/cockroachdb/cockroach/rpc/main_test.go:56 +0x28
		main.main()
			github.com/cockroachdb/cockroach/rpc/_test/_testmain.go:104 +0x1d1
FAIL
FAIL	github.com/cockroachdb/cockroach/rpc	0.980s
=== RUN TestAll
I0701 15:05:52.178879     184 rpc/codec/rpc_test.go:49] Arith.Add(1, 2): 3
I0701 15:05:52.179790     184 rpc/codec/rpc_test.go:49] Arith.Add(1, 2): 3
--- PASS: TestAll (0.00s)
PASS
ok  	github.com/cockroachdb/cockroach/rpc/codec	0.014s
?   	github.com/cockroachdb/cockroach/rpc/codec/message	[no test files]
?   	github.com/cockroachdb/cockroach/rpc/codec/wire	[no test files]
=== RUN TestGetCertificateUser
--- PASS: TestGetCertificateUser (0.00s)

```
Please assign, take a look and update the issue accordingly.

--- a/rpc/server.go
+++ b/rpc/server.go
@@ -153,13 +153,13 @@ func (s *Server) Serve(handler http.Handler) {
 			defer s.mu.Unlock()
 
 			switch state {
-			case http.StateNew:
+			case http.StateNew, http.StateHijacked:
 				if s.closed {
 					conn.Close()
 					return
 				}
 				s.activeConns[conn] = struct{}{}
-			case http.StateHijacked, http.StateClosed:
+			case http.StateClosed:
 				delete(s.activeConns, conn)
 			}
 		},

